receiver/dcgmreceiver/scraper.go

```diff
@@ -220,188 +177,180 @@ func discoverRequestedFields(config *Config) []string {
 	return requestedFields
 }
 
-func (s *dcgmScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
-	err := s.initClient()
-	if err != nil || s.client == nil {
-		return s.mb.Emit(), err
+func (s *dcgmScraper) run(ctx context.Context, metricsCh chan<- map[uint]deviceMetrics) error {
+	defer close(metricsCh)
+	for {
+		client, _ := s.initClient()
+		// Ignore the error; it's logged in initClient.
+		if client != nil {
+			s.runOnce(ctx, client, metricsCh)
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case metricsCh <- map[uint]deviceMetrics{}:
+			// Un-hang any scrapers waiting for data, since we currently have no metrics to offer.
+		case <-time.After(s.initRetryTime):
+		}
 	}
+	return nil
+}
 
-	s.settings.Logger.Sugar().Debug("Client created, collecting metrics")
-	deviceMetrics, err := s.client.collectDeviceMetrics()
-	if err != nil {
-		s.settings.Logger.Sugar().Warnf("Metrics not collected; err=%v", err)
-		return s.mb.Emit(), err
+func (s *dcgmScraper) runOnce(ctx context.Context, client *dcgmClient, metricsCh chan<- map[uint]deviceMetrics) {
+	defer client.cleanup()
+	for {
+		waitTime, err := client.collect()
```

Comment:
`collect()` returns the duration of data it scraped, as a hint for the next poll interval. Then we do the subsequent math on it, to produce the time we're going to wait. I reused the same variable but I suppose I could split that into two names.

I think this math is actually broken now, BTW, because we now pass `client.lastSuccessfulPoll` to `dcgm.GetValuesSince`. So if the first poll returns 10s of data, the math will compute a 5s wait, it will then wait 5s and poll again, at which point we'll get 5s of data and the math will compute a 2.5s wait, etc. I think we probably need to compute the expected wait time from `maxKeepSamples`/`updateFreqUs` instead.

Reply:
Maybe we just need to apply `/ 2` to `s.config.CollectionInterval`? I.e.:
```go
		waitTime = max(
			100*time.Millisecond,
			min(s.config.CollectionInterval/2, waitTime),
		)
```
?

Reply:
No, because that will still inch smaller over time (10s poll returns 9.9s of data, 9.9s poll returns 9.8s of data, etc.)

Also the very first poll always returns 0 data, so the first `waitTime` is going to be 100ms.

Reply:
I've updated the code to return `updateFreqUs * maxKeepSamples` as the buffer size, which with the current knobs means we'll collect every 5s.